src/NServiceBus.FileBasedRouting/XmlRoutingFileParser.cs

```
@@ -42,23 +45,32 @@ public IEnumerable<EndpointRoutingConfiguration> Parse(XDocument document)
             return configs;
         }
 
-        static IEnumerable<Type> GetAllMessageTypes(XElement endpointElement, string singular, string plural)
+        IEnumerable<Type> GetAllMessageTypes(XElement endpointElement, string singular, string plural)
         {
             var handles = endpointElement.Element("handles");
 
             var separatelyConfiguredCommands = handles
                 ?.Elements(singular)
                 .Select(e => SelectCommand(e.Attribute("type").Value))
+                .Where(e => e != null)
```

wouldn't this cause invalid parts of the configuration to be "silently" ignored?

---

It logs a `warn` so I think it should be enough.

---

it does and that's the point. In a centralized configuration, you have all the routes, even though those messages / types might not be available on that endpoint.

I also thought a `warn` log would be enough.

---

I think ideally (not sure if this is possible in a reliable way) for a given type `t` we should check if we know it's assembly. If not, we ignore such a type without logging warning. We would only warn if we know the assembly name is fine and we have it loaded but the type name is not recognized. 

We risk that a route with an invalid assembly name will be silently ignored but I think it might be better than logging `warn` for each unknown type if the message contracts are spread across many assemblies.

---

I'm on board with avoiding exceptions because the specific assembly isn't available on a given endpoint. But according to the linked issue, this would also silently ignore spelling errors although a given type might be available. The later scenario doesn't seem to be something we would want to swallow in my opinion, but at that point, I'm not sure how we could limit ourselves to the first scenario?

---

How about an option "I send messages from this assembly" declaration? By default we load all the types in each endpoints and log `warn` in the current code.

But if a given endpoint includes the list of assemblies it "knows", we ignore all routes for types from different assemblies. Yes, it adds a bit of effort in the multiple contract assembly case but it saves you `warn`s that are false positives.

e.g.
```
<endpoints>
  <endpoint name="endpointA">
    <contracts>
       <contract assembly="MyContracts"/>
    </contracts>
    <handles>
      <command type="Contracts.Commands.DemoCommand, Contracts" />
      <event type="Contracts.Events.DemoEvent, Contracts" />
    </handles>
  </endpoint>
  <endpoint name="endpointb">
```

---

@timbussmann I'm still wondering why a `warn` would not work in the case of an spelling issue? How would you validate the endpoint name? Is there way to ensure there's spelling mistake there? Why is the assembly/type different?

---

@SzymonPobiega hm that would work but I think an additional declaration adds quite a bit of pain to configure routing :( In that case I think I'd prefer going with @HEskandari suggested approach and log warnings.

@HEskandari I'm not aware of any validation regarding endpoint name. That's a good point. Although I think there's a bigger chance that you mispell some of the types and at least there we could help you by identifying that this type isn't around, it's a perfectly valid point that the assembly of a completely unrelated endpoint might not be around. Also, in case of commands, it should be less of an issue as send will also throw in case no route was found. So let's go ahead with your proposed changes 👍